services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceParentHelixAdmin.java

```diff
@@ -4801,6 +4801,81 @@ public Map<String, StoreDataAudit> getClusterStaleStores(String clusterName) {
     return retMap;
   }
 
+  /**
+   * This function iterates over all of Helix Parent Admin's child controllers,
+   * in order to obtain the stores and respective store information of each child controller,
+   * and then filter out the stores that are ready for compaction.
+   * @param clusterName, the name of the cluster to search for stores that are ready for compaction
+   * @return a list of StoreInfo objects of stores that are ready for compaction
+   */
+  @Override
+  public ArrayList<StoreInfo> getStoresForCompaction(String clusterName) {
+    ArrayList<StoreInfo> compactionReadyStores = new ArrayList<>();
+    try {
+      Map<String, ControllerClient> childControllers = getVeniceHelixAdmin().getControllerClientMap(clusterName);
+      ArrayList<StoreInfo> storeInfoList = new ArrayList<>();
+
+      // iterate through child controllers
+      for (Map.Entry<String, ControllerClient> controller: childControllers.entrySet()) {
+
+        // add all store info to storeInfoList
+        MultiStoreInfoResponse response = controller.getValue().getClusterStores(clusterName);
+        storeInfoList.addAll(response.getStoreInfoList());
+      }
+
+      // filter out
+      filterStoresForCompaction(storeInfoList, compactionReadyStores);
+    } catch (Exception e) {
+      throw new VeniceException("Something went wrong trying to fetch stores for compaction.", e);
+    }
+    return compactionReadyStores;
+  }
+
+  // package exclusive for testing
+  void filterStoresForCompaction(ArrayList<StoreInfo> storeInfoList, ArrayList<StoreInfo> compactionReadyStores) {
```

Comment:
Suggest that this returns the `compactionReadyStores` instead of mutating the input parameters.

It is a good practice to not mutate input parameters inside a function as it helps to reason about the changes to the variable. Keeping things immutable when possible makes it easy to reason about changes even when the code evolves.

Reply:
Yes I also agree and want to implement with this pattern. However, I need you help with some design decisions that I went through.

I made this choice because `getStoresForCompaction()` returns `compactionReadyStores` outside the `try` clause where `filterStoresForCompaction()` is invoked. Therefore, `compactionReadyStores` currently needs to be created outside the `try` clause.

What can we evolve to implement your suggested change?

Reply:
implemented